include/oneapi/dpl/pstl/hetero/dpcpp/execution_sycl_defs.h

```diff
@@ -42,21 +44,62 @@ struct DefaultKernelName;
 template <typename KernelName = DefaultKernelName>
 class device_policy
 {
+    // Needed for the copy constructor that rebinds the kernel name
+    template <typename>
+    friend class device_policy;
+
   public:
     using kernel_name = KernelName;
 
     device_policy() = default;
+
     template <typename OtherName>
-    device_policy(const device_policy<OtherName>& other) : q(other.queue())
+    device_policy(const device_policy<OtherName>& other)
+    {
+        ::std::scoped_lock lock{other.mtx};
+        q = other.q;
+    }
+
+    device_policy(const device_policy& other)
+    {
+        ::std::scoped_lock lock{other.mtx};
+        q = other.q;
+    }
+
+    device_policy(device_policy&& other)
     {
+        ::std::scoped_lock lock{other.mtx};
+        q.swap(other.q);
     }
+
+    device_policy&
+    operator=(const device_policy& other)
+    {
+        ::std::scoped_lock lock{mtx, other.mtx};
+        q = other.q;
+        return *this;
+    }
+
+    device_policy&
+    operator=(device_policy&& other)
+    {
+        ::std::scoped_lock lock{mtx, other.mtx};
+        q.swap(other.q);
+        return *this;
+    }
+
     explicit device_policy(sycl::queue q_) : q(q_) {}
-    explicit device_policy(sycl::device d_) : q(d_) {}
-    operator sycl::queue() const { return q; }
+    explicit device_policy(sycl::device d_) { q.emplace(d_); }
```

Comment:
Could you reorder the code so that these existing constructors come before the new copy & move ones, perhaps right after the line 54?

Reply:
Done.